optimum/habana/transformers/generation/utils.py

```diff
@@ -2128,6 +2151,8 @@ def expand_if_needed(tensor, new_size, value, dim=-1):
             inc = iter(incrementor(bucket_size, prompt_len))
         if bucket_size > 0:
             assert "position_ids" not in model_kwargs, "Untested path"
+        if self.generation_config.static_shapes:
+            initial_ids = input_ids[::num_beams, 0:cur_len]
```

Comment:
Just for my understanding, Is this slicing causing any bug?
I'm trying to understand as finalize_beams input shape will be changing now due to slice and will impact perf.

Reply:
I see the initial_ids.to("cpu") in     
sequence_outputs["sequences"] = finalize_beams(
                initial_ids.cpu(), move(beam_trace, "cpu"), self.config, self.generation_config.length_penalty
            )
and it's in "cpu", will such variable length cause perf too? for decode-only model, the initial_ids is a variable length tensor, you could test with the text-generation pipeline I provide in the review comments.

Reply:
I see 
    sequence_outputs["sequences"] = finalize_beams(
                initial_ids.cpu(), move(beam_trace, "cpu"), self.config, self.generation_config.length_penalty
            )
in the code. the initial_ids is in cpu when finalize_beams, will it also cause perf issue because of variable length?
for decode only case like text-generation, the initial_ids is a variable length tensor, you could test with the script I provide in the review comment

Reply:
@sywangyi I think having initial_ids on HPU with variable length cause perf degradation due to recompilation.
Moving to CPU might be better.

I will verify it once

Reply:
thanks